### PR TITLE
Register `<wunderbaum-tree>` component

### PR DESCRIPTION
**Pull request by mar10**

### CODE REVIEW

File: src/define_component.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
+import { Wunderbaum } from "./wunderbaum";
+
+function appendCssLink(url: string, shadowRoot?: ShadowRoot) {
+  let link = document.createElement("link");
+  link.setAttribute("rel", "stylesheet");
+  link.setAttribute("href", url);
+  if (shadowRoot) {
+    shadowRoot.appendChild(link);
```
Comment (jogibear9988) [inline src/define_component.ts:8]: Wouldn't it be better to include the style via css import and assign to components adoptet stylesheets?

File: src/define_component.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    const path = parts.join("/");
+    console.log(`Wunderbaum component using CSS path: ${path}`);
+
+    if (this.hasAttribute("icon-css")) {
+      appendCssLink(this.getAttribute("icon-css")!, shadowRoot);
+    }
+    appendCssLink(path, shadowRoot);
+
+    // @import url("${path}");
+    shadowRoot.appendChild(document.createElement("style")).textContent = `
```
Comment (jogibear9988) [inline src/define_component.ts:45]: also here, use adopted stylesheet?

Comment (jogibear9988) [reply src/define_component.ts:45]: you could use a helper like this: https://github.com/node-projects/base-custom-webcomponent/blob/master/src/BaseCustomWebComponent.ts#L28

then you could write the style in a nice way:

              style = css`
              :host {
              ....
              }`

Comment (jogibear9988) [reply src/define_component.ts:45]: andy you could create it outside of the constructor and reuse it

**Comment by mar10** [comment]
Thanks for your contribution.

I started this as an exploration, though I currently don't have a concrete use case for a WB web component,
So any kind of feedback is very welcome!
(Feel free to open a PR for your suggestions. We should avoid external dependencies though.)

I opened a [discussion](https://github.com/mar10/wunderbaum/discussions/33) for some general ideas.

**Comment by jogibear9988** [comment]
I don't wanted that you use this as an external lib, only wanted to show how I use the templateliteral.

I also don't need the component atm., but I need the style in a way so I could use it in my components.

I could work on the component if you'd like, but then we should allow many of the wunderbaum optionts to be set as properties.
A component with for example a property/attribute for the ajax datasource then could be used via dragdrop inside a designer without typeing one line of javascript

**Comment by jogibear9988** [comment]
I would do the component like this. 
Only a draft, there should be more of the options as properties... Will take a look wich ones are usefull

        import { Wunderbaum } from 'wunderbaum';
        //@ts-ignore
        import wunderbaumStyle from 'wunderbaum/dist/wunderbaum.css' assert { type: 'css' };
        
        const css = function (strings: TemplateStringsArray, ...values: any[]): CSSStyleSheet {
            const cssStyleSheet = new CSSStyleSheet();
            //@ts-ignore
            cssStyleSheet.replaceSync(toParString(strings, values));
            return cssStyleSheet;
        };
        
        let componentStyle = css` 
        :host {
          display: block;
          width: 100%;
          height: 100%;
        }`;
        
        const properties = ['selectMode']
        
        export class WunderbaumComponent extends HTMLElement {
            protected _tree?: Wunderbaum;
            protected _treeElement: HTMLDivElement;
        
            constructor() {
                super();
                const shadowRoot = this.attachShadow({ mode: "open" });
                shadowRoot.adoptedStyleSheets = [wunderbaumStyle, componentStyle];
        
                /*if (this.hasAttribute("icon-css")) {
                    appendCssLink(this.getAttribute("icon-css")!, shadowRoot);
                }*/
        
                this._treeElement = document.createElement("div");
                shadowRoot.appendChild(this._treeElement);
        
                //properties set before upgrade of component could override seter/geter from prototype
                for (let p of properties) {
                    if (this.hasOwnProperty(p)) {
                        const val = this[p];
                        delete this[p];
                        this[p] = val;
                    }
                }
            }
        
            static get observedAttributes() {
                return ["select-mode"];
            }
        
            attributeChangedCallback(name, oldValue, newValue) {
                switch (name) {
                    case 'select-mode':
                        this.selectMode = newValue;
                        break;
                }
            }
        
            connectedCallback() {
                this._tree = new Wunderbaum({
                    id: this.getAttribute("id")!,
                    element: this._treeElement,
                    header: this.getAttribute("header"),
                    selectMode: <any>this.getAttribute("select-mode") ?? "multi",
                    source: this.getAttribute("source") ?? [],
                    debugLevel: Number.parseInt(this.getAttribute("debug-level") ?? "3"),
                });
            }
        
            disconnectCallback() {
                this._tree?.destroy();
            }
        
            get tree(): Wunderbaum {
                if (!this._tree) {
                    throw new Error("Wunderbaum instance not yet initialized.");
                }
                return this._tree;
            }
        
            get source() {
                return this.getAttribute("source") || "";
            }
        
            get items() {
                return this.tree.toDictArray();
            }
        
            get selectMode() {
                return this.tree.options.selectMode;
            }
            set selectMode(value) {
                this.tree.options.selectMode = value;
            }
        }
        customElements.define("wunderbaum-tree", WunderbaumComponent);

**Pull request closed without merging**